### PR TITLE
[Gemma4] Replace one-hot matmul with F.embedding in position embeddings

### LINKED ISSUE #46175: [Gemma4] `Gemma4VisionPatchEmbedder._position_embeddings` materializes a ~19 GiB one-hot tensor that's mathematically a 2-row embedding lookup
### System Info

- `transformers` version: 5.6.2
- Platform: Linux 5.15.0-41-generic (x86_64)
- Python: 3.12
- PyTorch: 2.10.0+cu128
- GPU: 8 × NVIDIA H800 (80GB)
- Model: `google/gemma-4-E4B-it` (any Gemma 4 multimodal variant)
- Trainer: ZeRO-3 (DeepSpeed) full-parameter SFT, `torch_dtype: bfloat16`

### Who can help?

@yonigozlan @molbap (vision models) @zucchini-nlp (multimodal models)

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

The current implementation of `Gemma4VisionPatchEmbedder._position_embeddings` in
`src/transformers/models/gemma4/modeling_gemma4.py` (line 561 in 5.6.2) computes
2D patch position embeddings via one-hot encoding followed by a batched matmul:

```python
def _position_embeddings(self, pixel_position_ids, padding_positions):
    """Prepare patch positions map for matmul with positon embedding table."""
    clamped_positions = pixel_position_ids.clamp(min=0)
    one_hot = F.one_hot(clamped_positions, num_classes=self.position_embedding_size)
    one_hot = one_hot.permute(0, 2, 1, 3).to(self.position_embedding_table)
    position_embeddings = one_hot @ self.position_embedding_table
    position_embeddings = position_embeddings.sum(dim=1)
    position_embeddings = torch.where(padding_positions.unsqueeze(-1), 0.0, position_embeddings)
    return position_embeddings
```

For Gemma-4-E4B-it (`position_embedding_size=10240`), with batch_size=4 ×
10 images/sample × 2520 patches/image (the default `max_soft_tokens=280 *
pooling_kernel_size^2=9`), this materializes:

- `one_hot`: `[40, 2520, 2, 10240]` int64 → **15.38 GiB**
- `one_hot.to(bf16)` cast copy: same shape, bf16 → **3.85 GiB**

Total ~19 GiB just for two embedding lookups. Captured via
`torch.cuda.memory._record_memory_history` peak breakdown:

```
[1] alloc peak 36.00 GiB 
      19.57 GiB  (54.4%)  x10    vision tower (vision_model)
      14.09 GiB  (39.1%)        ZeRO-3 
       ...
```

Drilling into the 10 vision_model allocations:

```
one_hot @ line 581:   15.38 GiB (single int64 tensor, [40, 2520, 2, 10240])
one_hot.to bf16 @ 582: 3.85 GiB (bf16 cast of the same shape)
```

Stack tops:

```
_position_embeddings @ modeling_gemma4.py:581
forward @ modeling_gemma4.py:596
```

The `one_hot @ table` pattern is mathematically a 2-row embedding lookup
(two table rows indexed by `clamped[..., 0]` and `clamped[..., 1]`, summed)
that can be replaced with `F.embedding` without materializing any of the
`[..., position_embedding_size]` intermediates.

### Expected behavior

`_position_embeddings` should produce the same result without the
intermediate one-hot tensor. The following replacement is numerically
equivalent (bit-exact outside autocast; matches the original under autocast
to within bf16-matmul roundoff ~1.5e-2 absolute) and eliminates both the
15.38 GiB int64 tensor and its 3.85 GiB bf16 cast:

```python
def _position_embeddings(self, pixel_position_ids, padding_positions):
    """Prepare patch positions map via direct embedding lookup.

    Mathematically equivalent to the original one_hot @ table + sum,
    but avoids the [batch, num_patches, 2, position_embedding_size]
    one-hot temporary (~15 GiB int64 + ~3.85 GiB bf16 cast at bs=4,
    n_img=10, position_embedding_size=10240).
    """
    # Original .clamp(min=0) only bounds the lower side; one_hot() did
    # an implicit upper-bound check. F.embedding would read OOB on
    # overflow, so clamp both sides.
    clamped = pixel_position_ids.clamp(0, self.position_embedding_size - 1)
    x_emb = F.embedding(clamped[..., 0], self.position_embedding_table[0])
    y_emb = F.embedding(clamped[..., 1], self.position_embedding_table[1])
    position_embeddings = x_emb + y_emb
    # The original's trailing .sum(dim=1) promotes bf16 -> fp32 under
    # autocast (reduction-dtype rule); plain `+` doesn't, so cast
    # explicitly only inside autocast to keep the downstream
    # `hidden_states + pos` dtype contract.
    if torch.is_autocast_enabled() and position_embeddings.dtype != torch.float32:
        position_embeddings = position_embeddings.to(torch.float32)
    position_embeddings = torch.where(padding_positions.unsqueeze(-1), 0.0, position_embeddings)
    return position_embeddings
```

### PR DESCRIPTION
**Pull request by Sriniketh24**
## Summary

Fixes #46175

`Gemma4VisionPatchEmbedder._position_embeddings` materializes a one-hot tensor of shape `[batch, num_patches, 2, position_embedding_size]` in int64, then casts to the table's dtype, and matrix-multiplies against `position_embedding_table`. For typical training configs (`position_embedding_size=10240`, batch=40, 2520 patches), this allocates **~19 GiB** of GPU memory for what is mathematically a 2-row embedding lookup.

This PR replaces the `F.one_hot` + matmul pattern with two `F.embedding` calls (one per spatial axis), summed. The change:

- Eliminates the 15.38 GiB int64 one-hot tensor and its 3.85 GiB bf16 cast copy
- Preserves numerical equivalence (embedding lookup is the same operation)
- Clamps position IDs to `[0, position_embedding_size - 1]` (the original only clamped `min=0`; `F.embedding` would crash on OOB unlike `F.one_hot` which silently handled it)

### Coordination

- Issue discussion and approval: https://github.com/huggingface/transformers/issues/46175#issuecomment-4527166409
- Explicit approval from issue author @kuso2006: https://github.com/huggingface/transformers/issues/46175#issuecomment-4527861498
- No duplicate PRs exist (checked via `gh pr list --search "46175 in:body"`)

### Test

```bash
python -m pytest tests/models/gemma4/test_modeling_gemma4.py::Gemma4VisionPatchEmbedderTest -xvs
```

```
PASSED test_no_one_hot_intermediate
PASSED test_padding_zeroed
PASSED test_negative_positions_clamped
PASSED test_oob_positions_clamped
4 passed in 18.11s
```

AI-assisted contribution (Claude Code).

### CODE REVIEW

File: src/transformers/models/gemma4/modeling_gemma4.py [hunk 1]
```diff
@@ -584,15 +584,20 @@ def __init__(self, config: Gemma4VisionConfig):
...
-        position_embeddings = position_embeddings.sum(dim=1)
-        # Zero out embeddings for any padding patches.
+        """Compute 2-D patch position embeddings via embedding lookup.
+
+        ``pixel_position_ids`` has shape ``(batch, num_patches, 2)`` where the
+        last dimension holds (x, y) indices into ``position_embedding_table``
+        (shape ``(2, position_embedding_size, hidden_size)``).  The result is the
+        sum of the x- and y-embeddings for each patch.
+        """
+        clamped_positions = pixel_position_ids.clamp(min=0, max=self.position_embedding_size - 1)
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modeling_gemma4.py]: i am not sure about clamping the upper bound, `F.one_hot` raises an error when the values are beyond the total number of classes

Can you check what exactly happened, and why we need the upper bound clamping?

File: src/transformers/models/gemma4/modeling_gemma4.py [hunk 2]
```diff
@@ -584,15 +584,20 @@ def __init__(self, config: Gemma4VisionConfig):
...
+        # position_embedding_table: (2, position_embedding_size, hidden_size)
+        # clamped_positions[..., 0]: (batch, num_patches) — x indices
+        # clamped_positions[..., 1]: (batch, num_patches) — y indices
+        x_emb = F.embedding(clamped_positions[..., 0], self.position_embedding_table[0])
+        y_emb = F.embedding(clamped_positions[..., 1], self.position_embedding_table[1])
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modeling_gemma4.py:603]: yep, lgtm!

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 3]
```diff
@@ -863,3 +863,55 @@ def test_export_text_only(self):
...
         eager_generated_text = tokenizer.decode(eager_outputs[0], skip_special_tokens=True)
         self.assertEqual(export_generated_text, eager_generated_text)
+
+
+@require_torch
+class Gemma4VisionPatchEmbedderTest(unittest.TestCase):
+    """Unit tests for Gemma4VisionPatchEmbedder._position_embeddings."""
+
+    def _make_embedder(self, position_embedding_size=64, hidden_size=32):
+        from transformers import Gemma4VisionConfig
```
Comment (zucchini-nlp) [inline tests/models/gemma4/test_modeling_gemma4.py]: to remove, we don't need a test imo. I will trigger slow CI to check that model isn't broken

**Comment by zucchini-nlp** [comment]
run-slow: gemma4

File: src/transformers/models/gemma4/modeling_gemma4.py (hunk 1 above)
Comment (Sriniketh24) [reply src/transformers/models/gemma4/modeling_gemma4.py]: Good catch — removed the upper-bound clamp in the latest commit. The original code only had `clamp(min=0)` to handle negative padding sentinels, and `F.one_hot` would have raised on out-of-bounds values. I've matched that behavior: `clamp(min=0)` only, so valid positions remain in-range by construction and any out-of-bounds input would surface the same way it did before.

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 3 above)
Comment (Sriniketh24) [reply tests/models/gemma4/test_modeling_gemma4.py]: Done — removed the `Gemma4VisionPatchEmbedderTest` class entirely in the latest commit. Happy to let the slow CI integration tests cover this.

**Comment by zucchini-nlp** [comment]
run-slow: gemma4

**Review by zucchini-nlp — APPROVED** [review]
Merging, thanks for iterating!
